src/puppetlabs/jdbc_util/pool.clj

```diff
@@ -73,74 +73,81 @@
 (defn wrap-with-delayed-init
   "Wraps a connection pool that loops trying to get a connection, and then runs
   init-fn (with the connection as argument) before returning any connections to
-  the application. Accepts a timeout in ms that's used when deferencing the
+  the application. Accepts a timeout in ms that's used when dereferencing the
   future and by the status check. The datasource should have
   initialization-fail-fast set before being created or this is pointless."
-  [^HikariDataSource datasource init-fn timeout]
-  (when-not (.getHealthCheckRegistry datasource)
-    (.setHealthCheckRegistry datasource (HealthCheckRegistry.)))
-  (let [init-error (atom nil)
-        pool-future
-        (future
-          (loop []
-            (if-let [result
-                     (try
-                       ;; Try to get a connection to make sure the db is ready
-                       (.close (.getConnection datasource))
-                       (try (init-fn {:datasource datasource})
-                            (catch Exception e
-                              (swap! init-error (constantly e))
-                              (log/errorf e (trs "{0} - An error was encountered during initialization." (.getPoolName datasource)))))
-                       datasource
-                       (catch SQLTransientException e
-                         (log/warnf e (trs "{0} - Error while attempting to connect to database, retrying." (.getPoolName datasource)))
-                         nil))]
-              result
-              (recur))))]
-    (reify
-      DataSource
-      (getConnection [this]
-        (.getConnection (or (deref pool-future timeout nil)
-                            (throw (SQLTransientConnectionException. (tru "Timeout waiting for the database pool to become ready."))))))
-      (getConnection [this username password]
-        (.getConnection (or (deref pool-future timeout nil)
-                            (throw (SQLTransientConnectionException. (tru "Timeout waiting for the database pool to become ready."))))
-                        username
-                        password))
+  ([^HikariDataSource datasource init-fn timeout]
+   (wrap-with-delayed-init datasource init-fn datasource timeout))
+  ([^HikariDataSource datasource init-fn init-datasource timeout]
```

Comment:
We should document that init-datasource is a thing that works with java.jdbc (i.e. either a db map or a real datasource)